source/opt/local_access_chain_convert_pass.cpp

```diff
@@ -0,0 +1,366 @@
+  BuildAndAppendInst(
+      SpvOpCompositeInsert, varPteTypeId, insResultId, ins_in_opnds, newInsts);
+
+  // Build and append Store
+  BuildAndAppendInst(SpvOpStore, 0, 0, 
+      {{spv_operand_type_t::SPV_OPERAND_TYPE_ID, {varId}}, 
+       {spv_operand_type_t::SPV_OPERAND_TYPE_ID, {insResultId}}},
+      newInsts);
+}
+
+bool LocalAccessChainConvertPass::IsConstantIndexAccessChain(
+    ir::Instruction* acp) {
+  uint32_t inIdx = 0;
+  uint32_t nonConstCnt = 0;
+  acp->ForEachInId([&inIdx, &nonConstCnt, this](uint32_t* tid) {
+    if (inIdx > 0) {
+      ir::Instruction* opInst = def_use_mgr_->GetDef(*tid);
+      if (opInst->opcode() != SpvOpConstant) ++nonConstCnt;
+    }
+    ++inIdx;
+  });
+  return nonConstCnt == 0;
+}
+
+bool LocalAccessChainConvertPass::LocalAccessChainConvert(ir::Function* func) {
+  // Rule out variables accessed with non-constant indices
+  for (auto bi = func->begin(); bi != func->end(); ++bi) {
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+      switch (ii->opcode()) {
+      case SpvOpStore:
+      case SpvOpLoad: {
+        uint32_t varId;
+        ir::Instruction* ptrInst = GetPtr(&*ii, &varId);
+        if (ptrInst->opcode() != SpvOpAccessChain)
+          break;
+        if (!IsTargetVar(varId))
+          break;
+        if (!IsConstantIndexAccessChain(ptrInst)) {
+          seen_non_target_vars_.insert(varId);
+          seen_target_vars_.erase(varId);
+          break;
+        }
+      } break;
+      default:
+        break;
+      }
+    }
+  }
+  // Replace access chains of all targeted variables with equivalent
+  // extract and insert sequences
+  bool modified = false;
+  for (auto bi = func->begin(); bi != func->end(); ++bi) {
+    for (auto ii = bi->begin(); ii != bi->end(); ++ii) {
+      switch (ii->opcode()) {
+      case SpvOpLoad: {
+        uint32_t varId;
+        ir::Instruction* ptrInst = GetPtr(&*ii, &varId);
+        if (ptrInst->opcode() != SpvOpAccessChain)
+          break;
+        if (!IsTargetVar(varId))
+          break;
+        std::vector<std::unique_ptr<ir::Instruction>> newInsts;
+        uint32_t replId =
+            GenAccessChainLoadReplacement(ptrInst, &newInsts);
+        ReplaceAndDeleteLoad(&*ii, replId, ptrInst);
+        ++ii;
+        ii = ii.InsertBefore(&newInsts);
+        ++ii;
+        modified = true;
+      } break;
+      case SpvOpStore: {
+        uint32_t varId;
+        ir::Instruction* ptrInst = GetPtr(&*ii, &varId);
+        if (ptrInst->opcode() != SpvOpAccessChain)
+          break;
+        if (!IsTargetVar(varId))
+          break;
+        std::vector<std::unique_ptr<ir::Instruction>> newInsts;
+        uint32_t valId = ii->GetSingleWordInOperand(kSpvStoreValId);
+        GenAccessChainStoreReplacement(ptrInst, valId, &newInsts);
+        def_use_mgr_->KillInst(&*ii);
+        DeleteIfUseless(ptrInst);
+        ++ii;
+        ii = ii.InsertBefore(&newInsts);
+        ++ii;
+        ++ii;
```

Comment:
Yes :) Loads are replaced with two instructions: a Load and an Extract. Stores are replaced with three instructions: a Load, an Insert and a Store. Hence the extra ++ii.